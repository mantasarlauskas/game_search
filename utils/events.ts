import { MouseEvent, KeyboardEvent } from 'react';

export type ElementEvent = MouseEvent | KeyboardEvent;

export function onEnter(fn?: (e: ElementEvent) => void) {
    return (e: ElementEvent) => {
        if ((e as KeyboardEvent).key === 'Enter') {
            fn?.(e);
        }
    };
}
