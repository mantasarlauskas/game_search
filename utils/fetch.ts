export enum API_PATH {
    GAMES = 'games',
}

export async function fetchData(path: string, query?: string) {
    return fetch(
        `https://api.rawg.io/api/${path}?${query || ''}`,
    ).then((res) => {
        if (!res.ok) {
            return null;
        }

        return res.json();
    });
}
