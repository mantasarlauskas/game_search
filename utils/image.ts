export function cropImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}
