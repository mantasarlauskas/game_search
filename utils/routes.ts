export enum Route {
    HOME = '/',
    GAMES = '/games',
    GENRES = '/genres',
    PLATFORMS = '/platforms',
    NOT_FOUND = '/not-found',
    TAGS = '/tags'
}
