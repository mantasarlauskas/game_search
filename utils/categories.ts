import { API_PATH, fetchData, QueryParams } from 'utils/fetch';
import { Route } from 'utils/routes';
import { Category } from 'utils/types';

export function getCategoryTitle(categories: Category[], id: string) {
    return categories.find((category) => category.id === parseInt(id, 10))?.name;
}

export async function getCategoriesServerSideProps(id: string, pageSize: number, query: QueryParams) {
    const data = await fetchData(API_PATH.GAMES, {
        page: 1,
        page_size: pageSize,
        ...query,
    });

    if (!data?.count || !data?.results) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
        },
    };
}

export async function getCategoryAndGamesServerSideProps(
    path: API_PATH,
    id: string,
    pageSize: number,
    query: QueryParams,
) {
    const [data, category] = await Promise.all([
        fetchData(API_PATH.GAMES, {
            page: 1,
            page_size: pageSize,
            ...query,
        }),
        fetchData(`${path}/${id}`),
    ]);

    if (!data?.count || !data?.results || !category?.name) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
            name: category.name,
        },
    };
}
