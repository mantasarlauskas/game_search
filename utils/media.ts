import { Movie } from 'utils/types';

export function getCroppedImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}

export function getMovieUrl(movies: Movie[]) {
    return movies[Math.floor(Math.random() * movies.length)]?.data?.[480] || null;
}
