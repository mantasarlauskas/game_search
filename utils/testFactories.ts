import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
