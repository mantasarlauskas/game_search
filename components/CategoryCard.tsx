import Link from 'next/link';
import { Category } from 'utils/types';
import styles from 'components/CategoryCard.module.scss';
import { Route } from 'utils/routes';
import { getCroppedImageUrl } from 'utils/media';

function CategoryCard({
    route,
    category: { image_background, name, id, games_count, games },
    visibleGameCount = 3,
}: CategoryCardProps) {
    const imageUrl = getCroppedImageUrl(image_background);
    const backgroundImage = `linear-gradient(rgba(32, 32, 32, 0.5), rgb(32, 32, 32) 70%), url(${imageUrl})`;
    return (
        <div
            className={styles.root}
            style={{ backgroundImage }}
        >
            <Link href={`${route}/${id}`}>
                <div className={styles.name}>
                    {name}
                </div>
            </Link>
            <div className={styles.heading}>
                <div className={styles.title}>
                    Games
                </div>
                <div className={styles.count}>
                    {games_count}
                </div>
            </div>
            {games.slice(0, visibleGameCount).map((game) => (
                <Link key={game.slug} href={`${Route.GAMES}/${game.slug}`}>
                    <div className={styles.row}>
                        {game.name}
                    </div>
                </Link>
            ))}
        </div>
    );
}

interface CategoryCardProps {
    category: Category;
    visibleGameCount?: number;
    route: Route;
}

export default CategoryCard;
