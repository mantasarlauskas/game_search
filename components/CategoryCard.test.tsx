import { render, screen } from '@testing-library/react';
import CategoryCard from 'components/CategoryCard';
import { getCategories } from 'testing/testFactories';
import { Route } from 'utils/routes';
import userEvent from '@testing-library/user-event';

describe('<CategoryCard />', () => {
    const props = {
        route: Route.DEVELOPERS,
        category: getCategories()[0],
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('matches snapshot', () => {
        const { container } = render(<CategoryCard {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('shows name and games', () => {
        render(<CategoryCard {...props} />);
        expect(screen.getByText('Feral Interactive')).toBeInTheDocument();
        expect(screen.getByText('Rise of the Tomb Raider')).toBeInTheDocument();
        expect(screen.getByText('Middle-earth: Shadow of Mordor')).toBeInTheDocument();
        expect(screen.getByText('Hitman')).toBeInTheDocument();
    });

    it('on name click goes to category page', () => {
        render(<CategoryCard {...props} />);
        userEvent.click(screen.getByText('Feral Interactive'));
        expect(window.open).toBeCalledWith('/developers/18893');
    });

    it('on game click goes to game page', () => {
        render(<CategoryCard {...props} />);
        userEvent.click(screen.getByText('Hitman'));
        expect(window.open).toBeCalledWith('/games/hitman');
    });
});
