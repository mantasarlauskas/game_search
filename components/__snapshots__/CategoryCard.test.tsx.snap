// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryCard /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div>
      <div
        class="name"
      >
        Feral Interactive
      </div>
    </div>
    <div
      class="heading"
    >
      <div
        class="title"
      >
        Games
      </div>
      <div
        class="count"
      >
        107
      </div>
    </div>
    <div>
      <div
        class="row"
      >
        Rise of the Tomb Raider
      </div>
    </div>
    <div>
      <div
        class="row"
      >
        Middle-earth: Shadow of Mordor
      </div>
    </div>
    <div>
      <div
        class="row"
      >
        Hitman
      </div>
    </div>
  </div>
</div>
`;
